source/common/shared_pool/shared_pool.h

```diff
@@ -0,0 +1,116 @@
+ * control the life cycle of shared objects by reference counting
+ *
+ * Note:  ObjectSharedPool needs to be created in the main thread,
+ * all the member methods can only be called in the main thread,
+ * it does not have the ownership of object stored internally, the internal storage is weak_ptr,
+ * when the internal storage object destructor executes the custom deleter to remove its own
+ * weak_ptr from the ObjectSharedPool.
+ *
+ * There is also a need to ensure that the thread where ObjectSharedPool's destructor is also in the
+ * main thread, or that ObjectSharedPool destruct before the program exit
+ */
+template <typename T, typename HashFunc = std::hash<T>,
+          class = typename std::enable_if<std::is_copy_constructible<T>::value>::type>
+class ObjectSharedPool : public Singleton::Instance,
+                         public std::enable_shared_from_this<ObjectSharedPool<T, HashFunc>>,
+                         NonCopyable {
+public:
+  ObjectSharedPool(Event::Dispatcher& dispatcher)
+      : thread_id_(std::this_thread::get_id()), dispatcher_(dispatcher) {}
+
+  void deleteObject(const size_t hash_key) {
+    if (std::this_thread::get_id() == thread_id_) {
+      // There may be new inserts with the same hash value before deleting the old element,
+      // so there is no need to delete it at this time.
+      if (object_pool_.find(hash_key) != object_pool_.end() &&
+          object_pool_[hash_key].use_count() == 0) {
+        object_pool_.erase(hash_key);
+      }
+    } else {
+      // Most of the time, the object's destructor occurs in the main thread, but with some
+      // exceptions, it is destructed in the worker thread. In order to keep the object_pool_ thread
+      // safe, the deleteObject needs to be delivered to the main thread.
+      auto this_shared_ptr = this->shared_from_this();
+      // Used for testing to simulate some race condition scenarios
+      sync_.syncPoint(DeleteObjectOnMainThread);
+      dispatcher_.post([hash_key, this_shared_ptr] { this_shared_ptr->deleteObject(hash_key); });
+    }
+  }
+
+  std::shared_ptr<T> getObject(const T& obj) {
+    ASSERT(std::this_thread::get_id() == thread_id_);
+    auto hashed_value = HashFunc{}(obj);
+    auto object_it = object_pool_.find(hashed_value);
+    if (object_it != object_pool_.end()) {
+      auto lock_object = object_it->second.lock();
+      if (lock_object) {
+        return lock_object;
+      }
+    }
+
+    auto this_shared_ptr = this->shared_from_this();
+    std::shared_ptr<T> obj_shared(new T(obj), [hashed_value, this_shared_ptr](T* ptr) {
+      this_shared_ptr->sync().syncPoint(ObjectSharedPool<T>::ObjectDeleterEntry);
+      // release ptr as early as possible to avoid exposure of ptr, resulting in undefined behavior.
+      delete ptr;
+      this_shared_ptr->deleteObject(hashed_value);
+    });
+
+    // When inserted, it is possible that the old elements still exist before they can be deleted,
```

Comment:
Can you give an example sequence here that demonstrates?

Reply:
https://github.com/envoyproxy/envoy/blob/45d86449ff28395b3e9edf2db1b001e633f21190/test/common/shared_pool/shared_pool_test.cc#L120

 I don't know if this test case is enough for you to understand?

Reply:
I'm not grokking it. I think putting in code comments a better explanation/example would be helpful, thanks.